File: structsCPP/Stack/Stack.cpp
//
// Created by Angel Hernandez on 10/6/24.
//

#include "Stack.hpp"
#include <limits>


Stack::Stack() {
    this->top = nullptr;
}

Stack::Stack(const int &value) {
    Node* newNode = new Node(value);
    top = newNode;
    size = 1;
}

void Stack::push(const int &value) {
    Node* newNode = new Node(value);
    newNode->next = top;
    top = newNode;
    size++;
}

int Stack::pop() {
    if (size == 0) {
        return INT_MIN;
    }

    Node* temp = top;
    int popValue = temp->value;
    top = temp->next;
    delete temp;
    size--;
    return popValue;
}

void Stack::print() const {
    Node* temp = top;
    for (int i = 0; i < size; i++) {
        std::cout << temp->value << '\n';
        temp = temp->next;
    }
}

int Stack::getTop() const {
    return top->value;
}

int Stack::getSize() const {
    return size;
}

File: structsPYTHON/Stack/Stack/Stack.py
class Node:

    def __init__(self, value):
        self.value = value
        self.next = None


class Stack:

    def __init__(self, value):
        new_node = Node(value)
        self.top = new_node
        self.height = 1

    def push(self, value) -> None:
        new_node = Node(value)
        if self.is_empty() is not True:
            new_node.next = self.top

        self.top = new_node
        self.height += 1


    def pop(self) -> Node | None:
        if self.is_empty():
            return None

        removed_node = self.top
        if self.height == 1:
            self.top = None
        else:
            self.top = removed_node.next

        self.height -= 1
        return removed_node

    def is_empty(self) -> bool:
        if self.height == 0:
            return True
        return False

    def print(self) -> None:
        if self.is_empty():
            print("Stack is empty!")
            return

        temp = self.top
        while temp:
            print(temp.value)
            temp = temp.next

    def get_top(self) -> int:
        return self.top.value

    def get_height(self) -> int:
        return self.height
